src/metabase/api/advanced_computation.clj

```diff
@@ -55,6 +61,47 @@
   [[first-query & more-queries] context]
   (qp/process-query first-query (append-queries-context context more-queries)))
 
+(defn run-query
+  "Run the pivot query"
+  ([query]
+   (run-query query {:executed-by api/*current-user-id*}))
+
+  ([query info]
+   (qp.streaming/streaming-response [qp-context :api]
+     (qp.store/with-store
+       (let [main-breakout           (:breakout (:query query))
+             col-determination-query (-> query
+                                         ;; TODO: move this to a bitmask or something that scales better / easier to use
+                                         (assoc-in [:query :expressions] {"pivot-grouping" [:ltrim (json/generate-string main-breakout)]})
+                                         (assoc-in [:query :fields] [[:expression "pivot-grouping"]]))
+             all-expected-cols       (qp/query->expected-cols (qp/query->preprocessed col-determination-query))
+             all-queries             (pivot/generate-queries query)
+             first-query             (first all-queries)
+             rest-queries            (rest all-queries)]
+         (process-multiple-queries
+          (concat [first-query] rest-queries)
+          (assoc qp-context
+                 :info (assoc info :context qp-context)
+                 ;; this function needs to be executed at the start of every new query to
+                 ;; determine the mapping for maintaining query shape
+                 :column-mapping-fn (fn [query]
+                                      (let [query-cols (map-indexed vector (qp/query->expected-cols (qp/query->preprocessed query)))]
```

Comment:
Have you considered making `query-cols` a map from name->index? I think it would be more clear than piecing out all the bits from vectors as below.

Reply:
Maybe? I can give that a shot.